### PR TITLE
mDNS-SD publishes correctly

### LINKED ISSUE #52: Rust SDK service discovery not working correctly
I've been working on a Zeroconf adaptor for a different project and rediscovered WebThings at the weekend. I thought I'd combine the 2 projects and show my adaptor discovering WebThings adaptors. Unfortunately, the Rust [example](https://github.com/WebThingsIO/webthing-rust/blob/master/examples/multiple-things.rs) does not appear to work correctly.

When it didn't work, I tried the Python SDK and this works as expected. Python uses [`zeroconf`](https://pypi.org/project/zeroconf/) but I'm unable to determine which Zeroconf crate the Rust SDK uses.

When I run the Rust example, the services are advertized but only briefly:

```bash
RUST_LOG=debug cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.09s
     Running `target/debug/rust-webthings`
[main] Starting server
[2020-12-21T19:04:00Z INFO  actix_server::builder] Starting 8 workers
[2020-12-21T19:04:00Z INFO  actix_server::builder] Starting "actix-web-service-0.0.0.0:8888" service on 0.0.0.0:8888
setting new humidity level: 16.848787307739258
setting new humidity level: 11.045598983764648
setting new humidity level: 2.723759889602661
setting new humidity level: 18.06167984008789
setting new humidity level: 18.948495864868164
setting new humidity level: 1.1590396165847778
setting new humidity level: 3.7067339420318604
setting new humidity level: 0.8639483451843262
setting new humidity level: 20.823877334594727
setting new humidity level: 18.774051666259766
setting new humidity level: 3.289645195007324
```
And:
```bash
avahi-browse --all
+ enp5s0 IPv6 LightAndTempDevice                            _webthing._tcp       local
+ wlp6s0 IPv6 LightAndTempDevice                            _webthing._tcp       local
- enp5s0 IPv6 LightAndTempDevice                            _webthing._tcp       local
- wlp6s0 IPv6 LightAndTempDevice                            _webthing._tcp       local
```

> **NOTE** The `+` indicate the service being advertised and the `-`, the services being withdrawn. This happens quickly (within a second)

The humidity levels are published after the service is deleted (`-`).

I can add the service manually to the gateway but it isn't discovered (using mDNS|SD) and I suspect it's because it isn't being advertised (for very long).

I can, of course, browse the device (`http://localhost:8888`) too:

```console
[2020-12-21T19:04:12Z INFO  actix_web::middleware::logger] 127.0.0.1:46348 "GET / HTTP/1.1" 200 2189 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.88 Safari/537.36" 0.002946
```

And:

```YAML
[{
  "@context": "https://iot.mozilla.org/schemas",
  "@type": ["OnOffSwitch", "Light"],
  "actions": {
    "fade": {
      "description": "Fade the lamp to a given level",
      "input": {
        "properties": {
          "brightness": {
            "maximum": 100,
            "minimum": 0,
            "type": "integer",
            "unit": "percent"
          },
          "duration": {
            "minimum": 1,
            "type": "integer",
            "unit": "milliseconds"
          }
        },
        "required": ["brightness", "duration"],
        "type": "object"
      },
      "links": [{
        "href": "/0/actions/fade",
        "rel": "action"
      }],
      "title": "Fade"
    }
  },
  "base": "http://localhost:8888/0",
  "description": "A web connected lamp",
  "events": {
    "overheated": {
      "description": "The lamp has exceeded its safe operating temperature",
      "links": [{
        "href": "/0/events/overheated",
        "rel": "event"
      }],
      "type": "number",
      "unit": "degree celsius"
    }
  },
  "href": "/0",
  "id": "urn:dev:ops:my-lamp-1234",
  "links": [{
    "href": "/0/properties",
    "rel": "properties"
  }, {
    "href": "/0/actions",
    "rel": "actions"
  }, {
    "href": "/0/events",
    "rel": "events"
  }, {
    "href": "ws://localhost:8888/0",
    "rel": "alternate"
  }],
  "properties": {
    "brightness": {
      "@type": "BrightnessProperty",
      "description": "The level of light from 0-100",
      "links": [{
        "href": "/0/properties/brightness",
        "rel": "property"
      }],
      "maximum": 100,
      "minimum": 0,
      "title": "Brightness",
      "type": "integer",
      "unit": "percent"
    },
    "on": {
      "@type": "OnOffProperty",
      "description": "Whether the lamp is turned on",
      "links": [{
        "href": "/0/properties/on",
        "rel": "property"
      }],
      "title": "On/Off",
      "type": "boolean"
    }
  },
  "security": "nosec_sc",
  "securityDefinitions": {
    "nosec_sc": {
      "scheme": "nosec"
    }
  },
  "title": "My Lamp"
}, {
  "@context": "https://iot.mozilla.org/schemas",
  "@type": ["MultiLevelSensor"],
  "actions": {},
  "base": "http://localhost:8888/1",
  "description": "A web connected humidity sensor",
  "events": {},
  "href": "/1",
  "id": "urn:dev:ops:my-humidity-sensor-1234",
  "links": [{
    "href": "/1/properties",
    "rel": "properties"
  }, {
    "href": "/1/actions",
    "rel": "actions"
  }, {
    "href": "/1/events",
    "rel": "events"
  }, {
    "href": "ws://localhost:8888/1",
    "rel": "alternate"
  }],
  "properties": {
    "level": {
      "@type": "LevelProperty",
      "description": "The current humidity in %",
      "links": [{
        "href": "/1/properties/level",
        "rel": "property"
      }],
      "maximum": 100,
      "minimum": 0,
      "readOnly": true,
      "title": "Humidity",
      "type": "number",
      "unit": "percent"
    }
  },
  "security": "nosec_sc",
  "securityDefinitions": {
    "nosec_sc": {
      "scheme": "nosec"
    }
  },
  "title": "My Humidity Sensor"
}]
```

### PR DESCRIPTION
**Pull request by DazWilkin**
mDNS Service Discovery (mDNS-SD) now works correctly.

Fixed: #52 

It's not easy to provide Rust-based unit test for this (see #53).

However, the service publication can be shown to work using the gateway:

```bash
docker run \
--name webthings-gateway \
--rm --interactive --tty \
--env=TZ=America/Los_Angeles \
--volume=${PWD}/webthings:/home/node/.webthings \
--network=host \
--log-opt max-size=1m \
--log-opt max-file=10 \
webthingsio/gateway:latest
```

Then browsing the Gateway's UI:

image

Ensure the "Web Thing: Native web thing support" addon is enabled:

image

Then, run the Rust example:

```bash
cargo run
[main] Starting server
setting new humidity level: 0.12784115970134735
setting new humidity level: 22.60777473449707
setting new humidity level: 30.48933982849121
setting new humidity level: 16.809547424316406
setting new humidity level: 14.687156677246094
```

Then, scan for devices:

image

> **NOTE** I renamed the devices to be prefixed `Rusty` to confirm this was working.

Alternatively:

```bash
avahi-browse --all | grep _webthing._tcp
+ wlp6s0 IPv6 RustyDevice                                   _webthing._tcp       local
+ enp5s0 IPv6 RustyDevice                                   _webthing._tcp       local
+ enp5s0 IPv4 RustyDevice                                   _webthing._tcp       local
```

> **NOTE** `avahi-browse` finds `ThingsType::Multiple` by the multiple's name. But (see below), will find `ThingsType::Single` by the device name, e.g.:

```bash
- wlp6s0 IPv6 RustyDevice                                   _webthing._tcp       local
- enp5s0 IPv4 RustyDevice                                   _webthing._tcp       local
- enp5s0 IPv6 RustyDevice                                   _webthing._tcp       local
```

### CODE REVIEW

File: src/server.rs [hunk 1]
```diff
@@ -908,13 +911,15 @@ impl WebThingServer {
...
     /// * `hostname` - optional host name, i.e. mything.com
+    /// * `dns_service` - optional libmdns::Service name
     /// * `ssl_options` - tuple of SSL options to pass to the actix web server
     /// * `action_generator` - action generator struct
     /// * `base_path` - base URL to use, rather than '/'
     pub fn new(
         things: ThingsType,
         port: Option<u16>,
         hostname: Option<String>,
+        dns_service: Option<libmdns::Service>,
```
Comment (mrstegeman) [inline src/server.rs:922]: What's the point of passing this parameter in if it's just overwritten on line 1130?

File: src/server.rs [hunk 2]
```diff
@@ -1121,17 +1127,12 @@ impl WebThingServer {
...
+            Some(responder.register(SERVICE_TYPE.to_owned(), name.clone(), port, &["path=/"]));
 
         #[cfg(feature = "ssl")]
         match self.ssl_options {
             Some(ref o) => {
-                responder.register(
-                    "_webthing._tcp".to_owned(),
-                    name.clone(),
-                    port,
-                    &["path=/", "tls=1"],
```
Comment (mrstegeman) [inline src/server.rs:1132]: The difference here that is now omitted is the `tls=1` key/value pair.

**Review by mrstegeman — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,12 @@
 
 ## [Unreleased]
 
+## [0.13.2] - 2020-12-27
+### Changed
+- Added libmdns::Service property called `dns_service` to WebThingServer
+- WebThingServer::new() now takes additional additional `dns_service` property which can be `None`
```
[the remark this answers is not in the document]
Comment (DazWilkin) [reply CHANGELOG.md]: Fixed: https://github.com/WebThingsIO/webthing-rust/pull/54/commits/7475a260a69ba326147f0776548998fe2e0bccf5

File: src/server.rs (hunk 1 above)
Comment (DazWilkin) [reply src/server.rs]: Fixed: https://github.com/WebThingsIO/webthing-rust/pull/54/commits/dc9056247f5e904eccef600b30da5a4049a7bd0d

File: src/server.rs (hunk 2 above)
Comment (DazWilkin) [reply src/server.rs:1132]: I missed that, apologies! I wondered why the line was duplicated ;-)

Fixed: https://github.com/WebThingsIO/webthing-rust/pull/54/commits/1a8c3170cf3dd71ac2560a57376801ce4309b966

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,12 @@
 
 ## [Unreleased]
 
+## [0.13.2] - 2020-12-27
+### Changed
+- Added libmdns::Service property called `dns_service` to WebThingServer
+- WebThingServer::new() now takes an additional property (`dns_service`) which can be `None`
```
Comment (mrstegeman) [inline CHANGELOG.md:8]: This is no longer true, so it can be removed.

**Review by mrstegeman — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 4 above)
Comment (DazWilkin) [reply CHANGELOG.md]: Fixed: https://github.com/WebThingsIO/webthing-rust/pull/54/commits/233fab92055d26ee577354682873b3f6beb0ff93

**Review by mrstegeman — APPROVED** [review]

**Pull request merged**